Disappearing messages

yougotwill opened the pull request:
- Updated right panel in a conversation to use overlay logic similar to the left panel.
- Improved design for disappearing messages
- Added panel radio buttons that are used for the disappearing message timer options
- Improved Conversation Header that can switch between the notification settings, disappearing message setting and group members. When you click on the heading it opens the related right panel option (if available)
- Added 2 new disappearing message modes (The original mode aka. Legacy will be removed in the future but will be kept for now for backwards compatibility / migration purposes)
  1. Disappear After Read
  2. Disappear After Send
- Added disappearing support to other visible messages (call, group update, media extraction notifications etc.)
- Use clock svg for disappearing countdown in all cases to keep in line with other platforms.
- Disappearing messages are now correctly deleted from the swarm and won't appear when restoring a device
- Users with newer desktop clients will get notified if they are sent messages from an older (legacy) client encouraging them to tell the other user to upgrade
- Attachments should be correctly deleted off disk (and off swarm) once deleted in the client (solves https://github.com/oxen-io/session-desktop-temp/issues/315, https://github.com/oxen-io/session-desktop-temp/issues/327)
- Added a release feature utility which is used to activate a feature based on a set unix timestamp
- Improved spacing and line height on toast notification

**NOTE: Disappearing messages syncing will be finished after this branch is merged into the user config branch**

yougotwill committed:
fix: moved disappearing message types to utils

yougotwill committed:
feat: added expirationType to message model with correct typings

added to migration

yougotwill committed:
fix: correctly write to db when setting changes

yougotwill committed:
feat: added basic support for different disappearing message modes

added expirationType and lastDisappearingMessageChangeTimestamp to conversations table, left a few comments for future work

yougotwill committed:
fix: rebuilt disappearing messages header beause syncthing

yougotwill committed:
feat: added default timer values for different modes

yougotwill committed:
feat: added panel label for timer options

yougotwill committed:
feat: updated right panel in a conversation to use overlay logic in preparation for disappearing messages

added panel buttons specific to this UI

yougotwill committed:
fix: improved migration queries

yougotwill committed:
feat: added panel radio buttons that list timer options

yougotwill committed:
feat: DisappearingMessageType -> DisappearingMessageConversationType

this is more explicit and allows use to use DisappearingMessageType in the message model

yougotwill committed:
feat: removed DisappearingMessageMenuItem from conversation context menu

yougotwill committed:
feat: made right panel scrollable, added subtitle to panelradiobutton and xl margin option

yougotwill committed:
feat: added expirationType and expireTimer to content message and updated message pipeline

this includes sync messages and updating the protobuf, haven't tested closed groups

yougotwill committed:
fix: make sure to update state when the covnersation changes

yougotwill committed:
feat: added legacy mode timer notification copy

yougotwill committed:
fix: remove unneeded SyncedExpiries type from the Protobuf

yougotwill committed:
fix: repaired closed group disappearing messages

yougotwill committed:
feat: added function for calculating the expire start time

updated copy for setting change messages

yougotwill committed:
fix: cleanup

yougotwill committed:
feat: added disabled state to right panel components

yougotwill committed:
feat: removed old disappearing message display from header

show disappearing message screen when clicking the correct option

yougotwill committed:
feat: started consolidating send and receive dm logic.

updated in parts but haven't test what happens

yougotwill committed:
feat: block sending disappearing messages of the wrong type in different conversations,

improved disappear after send or read logic, disable legacy sending for now

yougotwill committed:
feat: added new togglable subtitle to header

yougotwill committed:
feat: moved PropsForExpiringMessage to state ducks

yougotwill committed:
feat: improved sender disappear after reading time

yougotwill committed:
fix: only log errors when syncing an expire timer message and not all sync messages

yougotwill committed:
fix: duplicate timer update messages

yougotwill committed:
feat: DataExtractionNotifications now expire

MessageRequestResponses should not expire, pass yarn ready

yougotwill committed:
feat: improved styling between states

yougotwill committed:
feat: slight improved timing of sent message expire start time

yougotwill committed:
feat: added notice banner component

show notice when conversation is in legacy mode, extracted some UI components from the ConversationHeader, updated featureReleaseTimestamp

yougotwill committed:
feat: attempted to improve the accuracy of the start timer for disappear after send

yougotwill committed:
feat: timer notifications expire correctly when disappearing after sending

yougotwill committed:
feat: extracted expiry logic from the GenericReadableMessage to a new component

This allows use to use the expiry logic for different visible messages not just generic ones

yougotwill committed:
feat: Call Notifications now expire properlly

include the direction in the expiration props, setting the start timer doesnt directly update the model

yougotwill committed:
fix: TimerNotification text is now centered

yougotwill committed:
feat: disappear after reading now works

yougotwill committed:
fix: left align PanelRadioButton text

yougotwill committed:
wip: added new legacy mode to panel, improved backwards compatibility

legacy mode support in conversation header, added typing for the expireUpdate, next is sending support

yougotwill committed:
fix: include off in available modes in call cases

yougotwill committed:
feat: disabled new disappearing message modes behind a timed feature release function

yougotwill committed:
feat: control available modes depending on the conversation

yougotwill committed:
feat: updated db migration to support legacy mode

yougotwill committed:
feat: fixed double expiration update message issue

started creating an inheritable expiration message class, improved start triggering logic

yougotwill committed:
feat: improved syncing of timers for disappear after send

yougotwill committed:
feat: updated timer notification design and copy

yougotwill committed:
fix: prevent duplicate expire timer updates while keeping backwards compatibility

yougotwill committed:
feat: added hasOutdated Client to convo model, use feature release check to process disappearing messages

trigger banner when receiving a legacy disappearing message via hasOutdatedClient

yougotwill committed:
refactor: extracted header title to its own component

yougotwill committed:
feat: fixed closed groups

improved setExpirationStartTimestamp logic

yougotwill committed:
feat: updated clients can send ExpirationTimerUpdateMessages to older clients

fixed syncing of ExpirationTimerUpdateMessages

yougotwill committed:
fix: resolved some yarn ready issues

yougotwill committed:
feat: updated NoticeBanner style

yougotwill committed:
feat: disappearing after send is not working correctly

yougotwill committed:
fix: improved backwards compatibility

yougotwill committed:
feat: group invitiations now disappear

yougotwill committed:
feat: disappearing messages are now correctly deleted from the swarm after they expire

yougotwill committed:
fix: signature verification now works with updated method for extend and shorten

yougotwill committed:
feat: improved conversation header title

yougotwill committed:
feat: added ability to reset the releaseFeature logic by updating the featureReleaseTimestamp

yougotwill committed:
feat: added single mode view for closed groups and note to self

yougotwill committed:
feat: only show banner after v2 is released

yougotwill committed:
fix: reverted setToExpire timer,

fixed incorrect expirationLength value in getPropsForExpiringMessage

yougotwill committed:
fix: make sure expirationType and expireTimer are defined after v2 is released

yougotwill committed:
feat: only use timer svg for disappearing message

yougotwill committed:
feat: tested new expiration modes with legacy and new clients

fixed banner logic

yougotwill committed:
fix: read receipt messages now show expireTimer correctly

yougotwill committed:
fix: renamed ExpirationType enum value LEGACY to UNKNOWN

yougotwill committed:
fix: adding shorten and extend to expire rpc endpoint

yougotwill committed:
fix: typing animation alignment

yougotwill committed:
fix: make sure migration triggers with the correct date for current development

yougotwill committed:
feat: add comments after run through

yougotwill committed:
feat: conversationHeaderTitle is now keyboard accessible

SessionIconButton is now keyboard accessible if a tabIndex is set to > -1

yougotwill committed:
feat: moved setToExpire calls to the correct locations within the message pipeline depending on the expirationType of a message model

yougotwill committed:
feat: handle legacy disappearing messages more gracefully due to protobuf issues

added utility function for checking for undefined properties on a protobuf, renamed expireTimer to expirationTimer in some places

yougotwill committed:
fix: legacy message checks must look for the ExpirationTimerUpdate flag explicitly

use correct model for a message in a closed group

yougotwill committed:
fix: improved feature release logic

yougotwill committed:
fix: expire all outbound disapparing messages as soon as they are sent to match other platforms

yougotwill committed:
feat: use correct timer options and set the defaults

yougotwill committed:
feat: legacy mode now works for 1-1 chats

yougotwill committed:
fix: cleanup logging

yougotwill committed:
fix: improved logic for turning off disappearing messages

yougotwill committed:
fix: refactor and works for all legacy test cases, modern to modern is broken

yougotwill committed:
feat: renamed isLegacyMessage to isLegacyDataMessage

all works except modern to modern stage 2

yougotwill committed:
feat: verify disappearing messages arent supported in communities

yougotwill committed:
feat: updated hasOutdatedClient to track last user to send a legacy message

this makes the banner work properlly in closed groups

yougotwill committed:
feat: added dismiss button to NoticeBanner

yougotwill committed:
feat: made the conversation header subititle into its own component

yougotwill committed:
feat: fixed closed groups legacy mode

yougotwill committed:
fix: fixed timer alignment once and for all

yougotwill committed:
feat: improvements to the disabled state and group non-admins settings

separated components in OverlayDisappearingMessages

yougotwill committed:
feat: added proper note to self support

note to self is always delete after send

yougotwill committed:
fix: pass yarn ready

yougotwill committed:
fix: explicitly set expirationType to undefined if sending a legacy message

yougotwill committed:
fix: updated featureReleaseTimestamps for testing purposes

yougotwill committed:
fix: finally got all the test cases to pass!

removed misMatchedMessage and replaced with shouldDisappearButIsntMessage which is a more explicit check

yougotwill committed:
feat: ignore conversation settings when new modes are unlocked

yougotwill committed:
feat: added legacy to the ExpirationType enum in the protobuf and updated support

the desktop protobuf implementation doesnt support undefined for optional enum values this is a known bug.

yougotwill committed:
fix: fixed legacy disappearing messages on call notifications and others

yougotwill committed:
fix: fixed state of overlay options depending on release status

yougotwill committed:
fix: duplicate copy

yougotwill committed:
fix: resolve turning off disappearing messages empty messages in most cases

still requires more tweaking

yougotwill committed:
feat: reverted bak to stage 1

yougotwill committed:
feat: abstracted getAllValidOpenGroupV2ConversationRoomInfos function and handled the response correctly

just need to do testing

yougotwill committed:
fix: pass yarn ready

yougotwill committed:
feat: made progress on attachment deletions but can confirm 100%

yougotwill committed:
fix: iterate allRoominfos map correctly

testing complete

yougotwill committed:
feat: made sure to clear attachment properties after deletion

confirmed everything works through testing

yougotwill committed:
feat: added delete option to inbox request in sogs, linked with delete account modal

still needs testing and handling the response see TODO CONTINUE

Bilb commented on ts/components/conversation/right-panel/RightPanel.tsx:
```diff
@@ -0,0 +1,54 @@
+  const rightOverlayMode = useSelector(getRightOverlayMode);
+  const [showNewDisppearingMessageModes, setShowNewDisppearingMessageModes] = useState(false);
+
+  const checkForFeatureRelease = useCallback(async () => {
+    const isReleased = await checkIsFeatureReleased('Disappearing Messages V2');
+    return isReleased;
+  }, []);
+
+  useEffect(() => {
+    let isCancelled = false;
+    checkForFeatureRelease()
+      .then(result => {
+        if (isCancelled) {
+          return;
+        }
+        setShowNewDisppearingMessageModes(result);
+      })
+      .catch(() => {
+        if (isCancelled) {
+          return;
+        }
+      });
+
+    return () => {
+      isCancelled = true;
+    };
+  }, [checkForFeatureRelease]);
```
not sure what this is supposed to do, but that `useCallback` has no dependency so is always the same.
then the useEffect has just that `checkForFeatureRelease` dependency which is always the same too as got from the useCallback above.
This looks to me to just


```ts
  useEffect(() => {
    let isCancelled = false;

    checkIsFeatureReleased('Disappearing Messages V2')
      .then(result => {
        if (isCancelled) {
          return;
        }
        setShowNewDisppearingMessageModes(result);
      })
      .catch(() => {
        if (isCancelled) {
          return;
        }
      });

    return () => {
      isCancelled = true;
    };
  }, []);
```
no?

yougotwill replied:
I think you are correct. I probably wrote the `useCallback` out of habit to prevent linting errors. Will update

Bilb commented on ts/components/conversation/right-panel/overlay/disappearing-messages/TimeOptions.tsx:
```diff
@@ -0,0 +1,44 @@
+    <>
+      {!hasOnlyOneMode && <PanelLabel>{window.i18n('timer')}</PanelLabel>}
+      <PanelButtonGroup>
+        {options.map((option: any) => {
```
why the any? it's not typed right as a   `options: TimerOptionsArray | null;`?

yougotwill replied:
My mistake. It should be the `TimerOptionsEntry` type which I have now exported and imported.

Bilb commented on ts/components/dialog/DeleteAccountModal.tsx:
```diff
@@ -57,6 +59,16 @@ async function deleteEverythingAndNetworkData() {
+      // clear each inbox per sogs
+      for (const roomInfo of allRoomInfos.values()) {
+        // TODO CONTINUE testing - use a dummy account with some message requests and then if we restore from seed there should be no message requests.
+        await clearInbox(roomInfo);
```
should probably  catch any errors here so we still continue deleting the DB if just one of the inbox fails?
like what if a sogs is offline, you still want to be able to remove your account.

yougotwill replied:
Good point. Does this make sense?

```ts
    // clear all sogs inboxes (includes message requests)
    const allRoomInfos = await getAllValidOpenGroupV2ConversationRoomInfos();
    if (allRoomInfos && allRoomInfos.size > 0) {
      // clear each inbox per sogs
      for (const roomInfo of allRoomInfos.values()) {
        // TODO CONTINUE testing - use a dummy account with some message requests and then if we restore from seed there should be no message requests.
        try {
          await clearInbox(roomInfo);
        } catch (error) {
          window.log.info(
            `[deleteEverythingAndNetworkData]: Failed to clear inbox ${roomInfo}`,
            error
          );
          continue;
        }
      }
    }
```

Bilb replied:
yep that would work. I saw you updated the code with it already so closing this item.

Bilb commented on ts/hooks/useParamSelector.ts:
```diff
@@ -183,3 +183,61 @@ export function useMessageReactsPropsById(messageId?: string) {
+        return options.filter(option => {
+          return (
+            (hasOnlyOneMode && option.value === 0) ||
+            option.value === 10 || // 10 seconds (for development)
+            option.value === 30 || // 30 seconds (for development)
+            option.value === 60 || // 1 minute (for testing)
```
might be worth to make those dependent on process.env through a feature flag or something

yougotwill replied:
Should we use `NODE_APP_INSTANCE`?

Otherwise this could be a good time to add a `yarn dev`. Something like this?

```
"dev": "cross-env NODE_ENV=development NODE_APP_INSTANCE=devprod$MULTI electron ."
```

Bilb replied:
Let's keep it like this until merged with user config

yougotwill replied:
For now I have just commented them out for now and lest 1 minute for q/a

Bilb commented on ts/models/message.ts:
```diff
@@ -1178,16 +1272,27 @@ export class MessageModel extends Backbone.Model<MessageAttributes> {
       }
 
-      window?.log?.info('Set message expiration', {
+      window?.log?.info('WIP: Set message expiration', {
```
you can .debug those WIP logs if you want to keep them

yougotwill replied:
I've cleaned up the WIP logs and changed to .debug where appropriate. I left the WIP for the sync message one as a reminder that we still need to deal with it.

Bilb commented on ts/session/messages/outgoing/visibleMessage/ClosedGroupVisibleMessage.ts:
```diff
@@ -2,11 +2,14 @@ import { SignalService } from '../../../../protobuf';
-  identifier?: string;
-  groupId: string | PubKey;
+interface ClosedGroupVisibleMessageParams extends ClosedGroupMessageParams {
+  // TODO Do we need strings?
```
do we?

yougotwill replied:
So the answer to this is yes for this PR. There are a couple places where we use it as a string and it would require quite a bit of refactoring that should be done separately. See `ts/session/group/closed-group.ts` for more context. For now I'll update this comment to say refactor to PubKey only in future.

yougotwill replied:
We want it to be the `PubKey` only because the constructor verifies the string against the regex.

Bilb commented on ts/state/ducks/conversations.ts:
```diff
@@ -71,7 +75,17 @@ export type FindAndFormatContactType = {
+  isExpired?: boolean;
+};
+
+export interface PropsForExpirationTimer extends PropsForExpiringMessage {
```
you can also just use the & operator on type rather than making everything interfaces extending one another

yougotwill replied:
Nice catch and good point. I think I just use interface out of habit because that is how I first learned typescript inheritance. But using type + & is more consistent with the code base and better based on my mental model of typescript now thanks to type-level typescript. Fixed

Bilb commented on ts/util/releaseFeature.ts:
```diff
@@ -0,0 +1,67 @@
+    const oldIsFeatureReleased = (await Data.getItemById('featureReleased'))?.value;
+    // values do not exist in the db yet. Let's store false for now in the db and update our cached value.
+    if (oldIsFeatureReleased === undefined) {
+      await Data.createOrUpdateItem({ id: 'featureReleased', value: false });
```
just so you know, we'll need to start using Storage.put for those createOrUpdateItem most of the time, so we can keep track of them in redux.

yougotwill replied:
Good catch man! I forgot about the put method. Also I realised when testing this that we don't do any sort of migration in memory if we swap to V2 from Legacy. So I have added a fix (which we will remove later after V2 is released) in `OverlayDisappearingMessages.tsx`. Please give it a look for me when you have the chance? 🙏

yougotwill committed:
fix: remove unecessary useCallback for feature releaser

yougotwill committed:
fix: remove unecessary any type from TimerOptions

yougotwill committed:
fix: log but continue if there are errors while clearing out sogs inboxes

yougotwill committed:
fix: cleaned up WIP logging and used the window log debug function where appropriate

yougotwill commented on ts/node/migration/sessionMigrations.ts:
```diff
@@ -1204,6 +1205,71 @@ function updateToSessionSchemaVersion29(currentVersion: number, db: BetterSqlite
+      `ALTER TABLE ${CONVERSATIONS_TABLE} ADD COLUMN lastDisappearingMessageChangeTimestamp INTEGER DEFAULT 0;`
+    ).run();
+
+    db.prepare(`ALTER TABLE ${CONVERSATIONS_TABLE} ADD COLUMN hasOutdatedClient TEXT;`).run();
```
It needs to be a string to assist with closed groups because the banner must show the name of the user who sent the legacy message. This was an easy way to store that information.

yougotwill committed:
fix: updated comment for when to remove legacy code

yougotwill commented on ts/session/apis/open_group_api/sogsv3/sogsV3ClearInbox.ts:
```diff
@@ -0,0 +1,66 @@
+};
+
+export const clearInbox = async (roomInfos: OpenGroupRequestCommonType): Promise<boolean> => {
+  let success = false;
+
+  const converationId = getOpenGroupV2ConversationId(roomInfos.serverUrl, roomInfos.roomId);
+  const conversation = await Data.getConversationById(converationId);
+
+  if (!conversation) {
+    window.log.warn('clearInbox Matching conversation not found in db');
+  } else {
```
Good catch! I built up the habit somewhere to use the Data. methods. I'll try and remember to do in memory checks where it's appropriate. Fixed.

yougotwill committed:
fix: in clearinbox check in memory for conversation instead of db lookup

this is more performent

yougotwill commented on ts/session/apis/snode_api/SNodeAPI.ts:
```diff
@@ -554,6 +554,8 @@ export async function retrieveNextMessages(
     handleTimestampOffset('retrieve', json.t);
     await handleHardforkResult(json);
 
+    // console.log(`WIP: retrieveNextMessages`, json.messages);
```
Fixed

yougotwill commented on ts/session/apis/snode_api/expire.ts:
```diff
@@ -0,0 +1,302 @@
+  messageHashes: Array<string>
+) {
+  if (isEmpty(swarm)) {
+    throw Error(`WIP: expireOnNodes failed! ${messageHashes}`);
```
fixed

yougotwill committed:
fix: improved comment explaining how seen messages are expired

yougotwill commented on ts/session/apis/snode_api/swarmPolling.ts:
```diff
@@ -360,6 +360,8 @@ export class SwarmPolling {
     const newMessages = messages.filter((m: Message) => !dupHashes.includes(m.hash));
 
     if (newMessages.length) {
+      // TODO explain this better
```
Fixed

yougotwill commented on ts/session/messages/outgoing/ExpirableMessage.ts:
```diff
@@ -0,0 +1,66 @@
+          ? SignalService.Content.ExpirationType.UNKNOWN
+          : undefined,
+      // TODO could use isFinite?
+      expirationTimer: this.expireTimer && this.expireTimer > -1 ? this.expireTimer : undefined,
```
The problem is that we can't differentiate between when it's undefined or if the value is actually set to zero (when we turn off disappearing messages). I also get a type error with isFinite and doing Number(undefined) would just give us 0 anyway. So will remove the comment and leave it as is.

yougotwill committed:
fix: remove isFinite comment

yougotwill committed:
fix: updated groupId comment for future work

yougotwill commented on ts/session/sending/MessageSentHandler.ts:
```diff
@@ -40,7 +41,7 @@ async function handlePublicMessageSentSuccess(
 }
 
-// tslint:disable-next-line: cyclomatic-complexity
+// tslint:disable-next-line:// tslint:disable-next-line: cyclomatic-complexity
```
no idea I must have copy pasted badly. Fixed

yougotwill committed:
fix: clean disable-next-line on handleMessageSentSuccess

yougotwill committed:
fix: updated child types from PropsExpiringMessges

use type and the ampersand operator instead of the interface with extends for better consistency with the rest of the code base

yougotwill committed:
fix: use storage put instead of data methods so that we can track in redux

added a fix in OverlayDisappearingMessages so when v2 turns on an and the app is running a migration occurs

yougotwill committed:
fix: pass yarn ready

yougotwill committed:
fix: check if path exists for deleted attachment instead of trying to read that data

this is more performant since it should be deleted

yougotwill commented on ts/types/MessageAttachment.ts:
```diff
@@ -28,6 +28,17 @@ export const deleteExternalMessageFiles = async (message: {
+    try {
+      await Promise.all(
+        attachments.map(async (attachment: { path: string; thumbnail: any; screenshot: any }) => {
+          await readAttachmentData(attachment.path);
```
Good point! I've updated my code for this so you will need to give it a 2nd look. Fixed

yougotwill replied:
I moved the logic for checking if the path exists to`handleDiskDeletion()` inside of `ts/types/attachments/migrations.ts`.

yougotwill committed:
Merge branch 'pr_fixes' into disappearing_messages

yougotwill committed:
fix: improved disk deletion handling and pass yarn ready

need to use await with the pathExists function and cleaned up code into a handler function

Bilb commented on ts/util/expiringMessages.ts:
```diff
@@ -20,8 +49,17 @@ export async function destroyMessagesAndUpdateRedux(
+    // tslint:disable-next-line: prefer-for-of
+    for (let i = 0; i < messageIds.length; i++) {
+      const message = await Data.getMessageById(messageIds[i]);
+      await message?.cleanup();
```
no we can't. I am not too sure what would be a good fix here. We should probably send the messageIds to cleanup to the node side and do a single remove there, but depending on how slow that is, it will block the UI thread.

Bilb replied:
or include it in the `removeMessagesByIds` just below directly. But yeah, we don't have MessageModels and stuff like that on the node side, just a json dump. And as I said, it might cause issues with blocking the UI. Let's keep it the way you made it for now.

Bilb commented on ts/receiver/dataMessage.ts:
```diff
@@ -240,6 +243,15 @@ export async function handleSwarmDataMessage(
           sentAt: sentAtTimestamp,
         });
 
+  if (expireUpdate && !isEmpty(expireUpdate)) {
+    if (isSyncedMessage) {
+      // TODO handle sync messages expiring separately?
+      window.log.debug('WIP: Sync Message dropping');
+    } else {
+      msgModel = handleExpireUpdate(convoToAddMessageTo, msgModel, expireUpdate);
+    }
+  }
+
```
anything should be done for that TODO here ? 
      // TODO handle sync messages expiring separately?

yougotwill replied:
If my memory serves correctly this is where we would have used synced expiries before we changed our approach to user config. I think we need to make logic for handling disappearing message updates (using your branch) and then letting regular disappearing messages that are sync messages be processed by the pipeline as normal.

Bilb replied:
ok all good. We will check this once merged.
Can you just remove the uneeded  ` expireUpdate ` check in `  if (expireUpdate && !isEmpty(expireUpdate)) {`

yougotwill replied:
Fixed

yougotwill committed:
fix: updated disappearing messages protobuf content values

this prevents a clash with the userconfig properties to be merged in later

yougotwill committed:
fix: run expireUpdate handler on all messages and just return the model if there are no changes

yougotwill committed:
feat: comment out development timer options

yougotwill committed:
feat: added PropsForExpirationSettings type

yougotwill committed:
fix: added onclick function to subtitle

fixed opening correct settings option depending on header option displayed

yougotwill committed:
feat: added data test ids

yougotwill committed:
feat: improved robustness of changing the conversation header subtitle

the previous logic relied on the length of the subtitles array which didn't account for when it changed depending on the conversation we were on

yougotwill committed:
fix: remove panel-settings panel mode since it is not needed

yougotwill commented on ts/components/conversation/right-panel/RightPanel.tsx:
```diff
@@ -0,0 +1,49 @@
+    case 'disappearing-messages':
+      // TODO legacy messages support will be removed in a future release
+      return <OverlayDisappearingMessages unlockNewModes={showNewDisppearingMessageModes} />;
+    case 'panel-settings':
+    default:
+      return <OverlayRightPanelSettings />;
+  }
```
Fixed

yougotwill commented on ts/components/conversation/right-panel/overlay/disappearing-messages/TimeOptions.tsx:
```diff
@@ -0,0 +1,44 @@
+export const TimeOptions = (props: TimerOptionsProps) => {
+  const { options, selected, setSelected, hasOnlyOneMode, disabled } = props;
+
+  if (!options || isEmpty(options)) {
```
I don't think I can because typescript will give an error.

Bilb replied:
right, stupid typescript

yougotwill replied:
Turns out we need to update our lodash type definitions. This will resolve this issue but we need to update a few cases so will be done as a separate PR.

yougotwill commented on ts/components/conversation/message/message-item/GroupInvitation.tsx:
```diff
@@ -21,10 +21,15 @@ export const GroupInvitation = (props: PropsForGroupInvitation) => {
   return (
-    <ReadableMessage
+    <ExpirableReadableMessage
+      convoId={props.convoId}
       messageId={messageId}
+      direction={props.direction}
       receivedAt={receivedAt}
-      isUnread={isUnread}
+      isUnread={Boolean(isUnread)}
+      expirationLength={props.expirationLength}
+      expirationTimestamp={props.expirationTimestamp}
+      isExpired={props.isExpired}
```
Good idea! Fixed.

yougotwill commented on ts/components/conversation/message/message-item/GroupUpdateMessage.tsx:
```diff
@@ -71,16 +71,29 @@ const ChangeItem = (change: PropsForGroupUpdateType): string => {
       messageId={messageId}
       receivedAt={receivedAt}
       isUnread={isUnread}
+      direction={direction}
+      expirationLength={expirationLength}
+      expirationTimestamp={expirationTimestamp}
+      isExpired={isExpired}
```
Fixed

yougotwill commented on ts/components/conversation/message/message-item/notification-bubble/CallNotification.tsx:
```diff
@@ -54,17 +63,21 @@ export const CallNotification = (props: PropsForCallNotification) => {
 
   return (
-    <ReadableMessage
+    <ExpirableReadableMessage
       messageId={messageId}
       receivedAt={receivedAt}
+      direction={direction}
       isUnread={isUnread}
+      expirationLength={expirationLength}
+      expirationTimestamp={expirationTimestamp}
+      isExpired={isExpired}
```
Fixed

yougotwill committed:
feat: created getMessageExpirationProps selector and used in ExpirableReadableMessage

this stops us passing props down from the child components of ExpirableReadableMessage

yougotwill commented on ts/components/conversation/right-panel/overlay/disappearing-messages/TimeOptions.tsx:
```diff
@@ -0,0 +1,44 @@
+
+  return (
+    <>
+      {!hasOnlyOneMode && <PanelLabel>{window.i18n('timer')}</PanelLabel>}
```
We still need it in the parent OverlayDisappearingMessages.tsx so we will keep this as is.

yougotwill commented on ts/node/migration/sessionMigrations.ts:
```diff
@@ -1204,6 +1205,71 @@ function updateToSessionSchemaVersion29(currentVersion: number, db: BetterSqlite
+
+    db.prepare(`ALTER TABLE ${CONVERSATIONS_TABLE} ADD COLUMN hasOutdatedClient TEXT;`).run();
+
+    // TODO update to agreed value between platforms
+    const disappearingMessagesV2ReleaseTimestamp = 1706778000000; // unix 01/02/2024 09:00
+    // const disappearingMessagesV2ReleaseTimestamp = 1677488400000; // unix 27/02/2023 09:00
```
We use window.log.info for logging inside of releaseFeature.ts which means we can't use that option. I can swap it for console logic or do a check for if window is defined and otherwise use console log?

Bilb replied:
what I meant was to create a new file named releaseFeatureTimestamp.ts with just the exported timestamp in it.
You could then import that exported timestamp from the two different files just fine, even if one is using the window object and the other one not. I think?

yougotwill replied:
No worries. I have moved it into the constants file and future proofed it. Test and works

```
export const FEATURE_RELEASE_TIMESTAMPS: Record<string, number> = {
  // TODO update to agreed value between platforms
  DISAPPEARING_MESSAGES_V2: 1706778000000, // unix 01/02/2024 09:00
  // NOTE for testing purposes only
  // DISAPPEARING_MESSAGES_V2: 1677488400000, // unix 27/02/2023 09:00
};
```

yougotwill commented on ts/node/migration/sessionMigrations.ts:
```diff
@@ -1204,6 +1205,71 @@ function updateToSessionSchemaVersion29(currentVersion: number, db: BetterSqlite
+      db.prepare(
+        `UPDATE ${CONVERSATIONS_TABLE} SET
+      expirationType = $expirationType
+      WHERE type = 'group' AND is_medium_group = 1 AND expireTimer > 0;`
```
Added TODO

Bilb replied:
keeping as unresolved until merged to my PR

yougotwill commented on ts/node/sql.ts:
```diff
@@ -1454,6 +1471,7 @@ function getSeenMessagesByHashList(hashes: Array<string>) {
 function getExpiredMessages() {
   const now = Date.now();
 
+  // TODO probably need to update and also add functions to find specific kinds of disappearing messages
```
Yeah it's fine for now

yougotwill commented on ts/node/sql.ts:
```diff
@@ -1485,6 +1503,7 @@ function getOutgoingWithoutExpiresAt() {
   return map(rows, row => jsonToObject(row.json));
 }
 
+// TODO Maybe we need different queries for the different modes?
```
Yeah it's fine for now

yougotwill commented on ts/session/messages/outgoing/visibleMessage/ClosedGroupVisibleMessage.ts:
```diff
@@ -2,11 +2,14 @@ import { SignalService } from '../../../../protobuf';
-  identifier?: string;
-  groupId: string | PubKey;
+interface ClosedGroupVisibleMessageParams extends ClosedGroupMessageParams {
+  // TODO Refactor closed groups typings so groupId is PubKey only
```
It depends how much you are about the typings. Could be a quick bug breather task?

Bilb replied:
yeah let's keep the todo for now

yougotwill committed:
fix: cleanup comments

yougotwill committed:
fix: reverted code and updated comment on readReceipts pending qa

yougotwill commented on ts/util/readReceipts.ts:
```diff
@@ -45,21 +45,23 @@ async function onReadReceipt(receipt: { source: string; timestamp: number; readA
     });
 
-    if (message.isExpiring() && !expirationStartTimestamp) {
+    // I think this is redundent since expirationStartTimestamp is always undefined and this the function will null return
+    if (message.isExpiring() && expirationStartTimestamp) {
```
Have updated comment and reverted to clearnet version pending merge and qa.

Bilb replied:
ok keeping unresolved for now

yougotwill commented on ts/components/conversation/right-panel/overlay/disappearing-messages/OverlayDisappearingMessages.tsx:
```diff
@@ -0,0 +1,199 @@
+  // TODO legacy messages support will be removed in a future
+  useEffect(() => {
+    if (unlockNewModes && modeSelected === 'legacy' && selectedConversationKey) {
+      const convo = getConversationController().get(selectedConversationKey);
+      if (convo) {
```
We are updating the conversation expirationType in this function so we can't use the redux selected conversation

Bilb replied:
yeah all good. I thought this call was made in a selector but that's in a component. So all good

yougotwill committed:
fix: incorrect direction check in getPropsForExpiringMessage

yougotwill commented on ts/models/message.ts:
```diff
@@ -268,16 +278,43 @@ export class MessageModel extends Backbone.Model<MessageAttributes> {
+        ? expireTimerStart + expirationLength
+        : null;
+
+    const direction =
+      this.get('direction') || this.get('type') === 'outgoing' ? 'outgoing' : 'incoming';
```
Good catch!

yougotwill commented on ts/state/selectors/conversations.ts:
```diff
@@ -1174,3 +1178,67 @@ export const getIsSelectedConvoInitialLoadingInProgress = createSelector(
+export const getSelectedConversationExpirationModes = createSelector(
+  getSelectedConversation,
+  (convo: ReduxConversationType | undefined) => {
+    let modes = DisappearingMessageConversationSetting;
+    // TODO legacy messages support will be removed in a future release
+    // TODO remove legacy mode
+    modes = modes.slice(0, -1);
```
The conversation setting is a hardcoded array 
`export const DisappearingMessageConversationSetting = ['off', ...DisappearingMessageMode, 'legacy'];` 

We could do an indexOf and delete but since we will remove this code a few weeks after the release...

yougotwill commented on ts/state/selectors/conversations.ts:
```diff
@@ -1174,3 +1178,67 @@ export const getIsSelectedConvoInitialLoadingInProgress = createSelector(
+
+export const getSelectedConversationExpirationModes = createSelector(
+  getSelectedConversation,
+  (convo: ReduxConversationType | undefined) => {
+    let modes = DisappearingMessageConversationSetting;
+    // TODO legacy messages support will be removed in a future release
+    // TODO remove legacy mode
+    modes = modes.slice(0, -1);
+
+    // Note to Self and Closed Groups only support deleteAfterSend
+    const isClosedGroup = convo?.isGroup && !convo.isPublic;
+    if (convo?.isMe || isClosedGroup) {
+      modes = [modes[0], modes[2]];
+    }
```
Cool cool. Yeah will be easier to cleanup later

yougotwill committed:
feat: added DisappearingMessageModesWithState type

yougotwill commented on ts/state/selectors/conversations.ts:
```diff
@@ -1174,3 +1178,67 @@ export const getIsSelectedConvoInitialLoadingInProgress = createSelector(
+    // Legacy mode is the 2nd option in the UI
+    modes = [modes[0], modes[modes.length - 1], ...modes.slice(1, modes.length - 1)];
+
+    const modesWithDisabledState: any = {};
```
Done!

yougotwill commented on ts/state/selectors/conversations.ts:
```diff
@@ -1174,3 +1178,67 @@ export const getIsSelectedConvoInitialLoadingInProgress = createSelector(
+      modes = [modes[0], modes[2]];
+    }
+
+    const modesWithDisabledState: any = {};
```
Done!

yougotwill committed:
feat: print error message and not entire error if available when deleting an attachment from disk

yougotwill commented on ts/types/attachments/migrations.ts:
```diff
@@ -142,29 +143,49 @@ export const loadData = async (attachment: any) => {
+    window.log.debug(`deleteDataSuccessful: Deletion succeeded for attachment ${path}`);
+    return undefined;
+  } catch (err) {
+    window.log.warn(`deleteDataSuccessful: Deletion failed for attachment ${path} ${err}`);
```
done

yougotwill commented on ts/types/attachments/migrations.ts:
```diff
@@ -142,29 +143,49 @@ export const loadData = async (attachment: any) => {
-      await deleteOnDisk(thumbnail.path);
-    }
+  if (isString(path)) {
+    attachment.path = String(await handleDiskDeletion(path));
```
Good catch. Fixed should be `path && isString(path)` right?

Bilb replied:
is that what you meant?

yougotwill replied:
I see! Ended up having to change slightly because we can't assign attachment.path to undefined;

```
  if (path && isString(path)) {
    const pathAfterDelete = await handleDiskDeletion(path);
    if (isString(pathAfterDelete)) {
      attachment.path = pathAfterDelete;
    }
  }
```

yougotwill committed:
fix: check if path is undefined first in deleteData function

yougotwill committed:
fix: moved feature release timestamp to constants so it only has to be defined once

yougotwill committed:
fix: converting undefined to a string does not make an empty string so we have to make deleteData more robust

javascript ¯\_(ツ)_/¯

yougotwill committed:
feat: make timer value load an existing one from the conversation or the default

fixed session radio button flickering

yougotwill committed:
feat: improved spacing on toast notification

converted scss toast styling to styled components

yougotwill committed:
fix: disappearing modes loadDataTestId should return in switch cases

yougotwill commented on ts/components/conversation/right-panel/overlay/disappearing-messages/DisappearingModes.tsx:
```diff
@@ -3,6 +3,26 @@ import { DisappearingMessageConversationType } from '../../../../../util/expirin
 import { PanelButtonGroup, PanelLabel } from '../../../../buttons/PanelButton';
 import { PanelRadioButton } from '../../../../buttons/PanelRadioButton';
 
+function loadDataTestId(mode: DisappearingMessageConversationType) {
+  let dataTestId = 'disappear-%-option';
+  switch (mode) {
+    case 'legacy':
+      dataTestId = dataTestId.replace('%', 'legacy');
+      break;
+    case 'deleteAfterRead':
+      dataTestId = dataTestId.replace('%', 'after-read');
+      break;
+    case 'deleteAfterSend':
+      dataTestId = dataTestId.replace('%', 'after-send');
+      break;
+    case 'off':
+    default:
+      dataTestId = dataTestId.replace('%', 'off');
+  }
+
+  return dataTestId;
+}
+
```
done

yougotwill commented on ts/util/releaseFeature.ts:
```diff
@@ -32,7 +30,7 @@ export async function checkIsFeatureReleased(featureName: string): Promise<boole
 
     // Is it time to release the feature?
-    if (Date.now() >= featureReleaseTimestamp) {
+    if (Date.now() >= FEATURE_RELEASE_TIMESTAMPS[`${featureName}`]) {
```
I needed to do this because of moving the timestamp object to the constants file

```
export const FEATURE_RELEASE_TIMESTAMPS: Record<string, number> = {
  // TODO update to agreed value between platforms
  DISAPPEARING_MESSAGES_V2: 1706778000000, // unix 01/02/2024 09:00
  // NOTE for testing purposes only
  // DISAPPEARING_MESSAGES_V2: 1677488400000, // unix 27/02/2023 09:00
};
```

yougotwill commented on ts/components/conversation/right-panel/overlay/disappearing-messages/TimeOptions.tsx:
```diff
@@ -0,0 +1,45 @@
+              }}
+              noBackgroundColor={true}
+              disabled={disabled}
+              dataTestId={`disappear-time-${option.name.replace(' ', '-')}-option`}
```
I've seen if there is a way to generate an english version but we would need to add some functions in i18n so instead of chose to use the seconds value as the unique id. Then it can be used without worry about localisation.

yougotwill committed:
fix: changed timer options dataTestId to use seconds and not the name

since we need an id unaffected by localisation

yougotwill commented on ts/components/conversation/header/ConversationHeaderTitle.tsx:
```diff
@@ -0,0 +1,184 @@
+              {convoName}
+            </span>
+          )}
+          {subtitleArray.length && subtitleArray.indexOf(visibleSubtitle) > -1 && (
```
Very good point! I guess I never used length as a boolean check until recently so I never stumbled on this issue. Have removed it and we can rely on IndexOf. Fixed

yougotwill committed:
fix: checking the subtitle array without casting as bool will produce a string in react

yougotwill commented on ts/components/conversation/message/message-item/ExpirableReadableMessage.tsx:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    if (isExpired || now >= expirationTimestamp) {
+      await Data.removeMessage(messageId);
```
Discussed and we agreed it's fine. Resolved

yougotwill committed:
fix: mispell of ConversationHeaderSubtitle

yougotwill commented on ts/components/conversation/header/ConversationHeaderSubtitle.tsx:
```diff
@@ -0,0 +1,163 @@
+  showDisappearingMessageIcon: boolean;
+};
+
+export const ConversationHeaderSubitle = (props: ConversationHeaderSubitleProps) => {
```
Nice catch!  Fixed

yougotwill commented:
Closed in favour of https://github.com/oxen-io/session-desktop/pull/2940
